sdk/tracing/azotel/otel.go

```diff
@@ -0,0 +1,120 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// azotel adapts OpenTelemetry tracing for consumption by the azcore/tracing package.
+package azotel
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/tracing"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// TracingProviderOptions contains the optional values for NewTracingProvider.
+type TracingProviderOptions struct {
+	// for future expansion
+}
+
+// NewTracingProvider creates a new tracing.Provider that wraps the specified OpenTelemetry TracerProvider.
+//   - tracerProvider - the TracerProvider to wrap
+//   - opts - optional configuration. pass nil to accept the default values
+func NewTracingProvider(tracerProvider *tracesdk.TracerProvider, opts *TracingProviderOptions) tracing.Provider {
+	return tracing.NewProvider(func(namespace, version string) tracing.Tracer {
+		tracer := tracerProvider.Tracer(namespace, trace.WithInstrumentationVersion(version))
+
+		return tracing.NewTracer(func(ctx context.Context, spanName string, options *tracing.SpanOptions) (context.Context, tracing.Span) {
+			kind := tracing.SpanKindInternal
+			var attrs []attribute.KeyValue
+			if options != nil {
+				kind = options.Kind
+				attrs = convertAttributes(options.Attributes)
+			}
+			ctx, span := tracer.Start(ctx, spanName, trace.WithSpanKind(convertSpanKind(kind)), trace.WithAttributes(attrs...))
+			return ctx, convertSpan(span)
+		}, &tracing.TracerOptions{
+			SpanFromContext: func(ctx context.Context) (tracing.Span, bool) {
+				return convertSpan(trace.SpanFromContext(ctx)), true
+			},
+		})
+
+	}, nil)
+}
+
+func convertSpan(traceSpan trace.Span) tracing.Span {
+	impl := tracing.SpanImpl{
+		End: func() {
+			traceSpan.End()
+		},
+		SetAttributes: func(attrs ...tracing.Attribute) {
+			traceSpan.SetAttributes(convertAttributes(attrs)...)
+		},
+		AddEvent: func(name string, attrs ...tracing.Attribute) {
+			traceSpan.AddEvent(name, trace.WithAttributes(convertAttributes(attrs)...))
+		},
+		AddError: func(err error) {
+			traceSpan.RecordError(err)
+		},
+		SetStatus: func(code tracing.SpanStatus, desc string) {
+			traceSpan.SetStatus(convertStatus(code), desc)
+		},
+	}
+	return tracing.NewSpan(impl)
+}
+
+func convertAttributes(attrs []tracing.Attribute) []attribute.KeyValue {
+	if len(attrs) == 0 {
+		return nil
+	}
+	keyvals := []attribute.KeyValue{}
+	for _, kv := range attrs {
+		switch vv := kv.Value.(type) {
+		case int:
+			keyvals = append(keyvals, attribute.Int(kv.Key, vv))
+		case int64:
```

Comment:
Is there any "integer promotion" that needs to happen here for int32, uint, etc..?

Reply:
Good question.  I'm tempted to say no because OTel doesn't natively support other types (see [here](https://github.com/open-telemetry/opentelemetry-go/blob/main/attribute/kv.go)).  And we [document](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/tracing/tracing.go#L198) the natively supported types, falling back to `fmt.Sprintf()` as required.

@lmolkova do you have an opinion on this?

Reply:
We don't use these helpers externally.  Since `Value` is exposed as `any` type from `azcore`, we check the type here before calling the appropriate `attribute.Int()` etc helpers.  If we need to support other types that what we [documented](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/tracing/tracing.go#L197-L200) then we need to check for them here.

Reply:
Going to leave this as-is for now.  If callers don't want the `fmt.Sprintf()` fallback they can cast the values themselves (we can adjust accordingly per any feedback).